### ISSUE TITLE
nanomq server get stuck

### ISSUE THREAD

**Issue by mxchipwanges** [nanomq/nanomq#1666]
**Describe the bug**
nanomq server get stuck,  can not be connected by any mqtts/mqtt client.

**Expected behavior**
server can be connected by mqtt client.

**Actual Behavior**
Describe what occurred.

**To Reproduce**
If possible include actual reproduction test code here.
Minimal C test cases are perferred.
An android paho-mqttv3 client connect to the nanomq server twice at the same time (interval millisecond).

**Environment Details**
 - NanoMQ version   release 0.21.2
 - Operating system and version
 - Compiler and language used
 - testing scenario

**Client SDK**
An android paho-mqttv3 client

**Additional context**

/tmp/nanomq.log

2024-02-07 11:53:06 [1054] TRACE /home/runner/work/nanomq/nanomq/nng/src/sp/transport/mqtts/broker_tls.c:535: ############### tlstran_pipe_send_cb [0xae994350] ################
2024-02-07 11:53:06 [1054] TRACE /home/runner/work/nanomq/nanomq/nng/src/sp/transport/mqtts/broker_tls.c:549: tls over tcp socket sent 155 bytes iov 0
2024-02-07 11:53:06 [1054] TRACE /home/runner/work/nanomq/nanomq/nng/src/sp/transport/mqtts/broker_tls.c:1477: ########### tlstran_pipe_send_start ###########
2024-02-07 11:53:06 [1052] INFO  /home/runner/work/nanomq/nanomq/nng/src/sp/protocol/mqtt/nmq_mqtt.c:452: pipe 1853715459 occupied! resending in cb!
2024-02-07 11:53:06 [1052] TRACE /home/runner/work/nanomq/nanomq/nng/src/sp/protocol/mqtt/nmq_mqtt.c:375:  #### nano_ctx_send with ctx 0x633fb0 msg type 30 #### 
2024-02-07 11:53:06 [1055] DEBUG /home/runner/work/nanomq/nanomq/nng/src/sp/protocol/mqtt/nmq_mqtt.c:670: client connected! addr [192.168.31.176 port [29903]

2024-02-07 11:53:06 [1054] TRACE /home/runner/work/nanomq/nanomq/nng/src/sp/transport/mqtts/broker_tls.c:1487: aio not functioning
2024-02-07 11:53:06 [1054] TRACE /home/runner/work/nanomq/nanomq/nng/src/sp/protocol/mqtt/nmq_mqtt.c:946: ******** nano_pipe_send_cb 96152714 ****
2024-02-07 11:53:06 [1055] TRACE /home/runner/work/nanomq/nanomq/nng/src/sp/protocol/mqtt/mqtt_parser.c:1243: no valid entry in auth list
2024-02-07 11:53:06 [1054] DEBUG /home/runner/work/nanomq/nanomq/nanomq/apps/broker.c:631: SEND ^^^^ ctx4 ^^^^
2024-02-07 11:53:06 [1053] TRACE /home/runner/work/nanomq/nanomq/nng/src/sp/protocol/mqtt/nmq_mqtt.c:375:  #### nano_ctx_send with ctx 0x635378 msg type 30 #### 
2024-02-07 11:53:06 [1054] DEBUG /home/runner/work/nanomq/nanomq/nanomq/pub_handler.c:1286: free topic
2024-02-07 11:53:06 [1054] DEBUG /home/runner/work/nanomq/nanomq/nanomq/pub_handler.c:1302: free payload
2024-02-07 11:53:06 [1054] DEBUG /home/runner/work/nanomq/nanomq/nanomq/pub_handler.c:1308: free pub_packet
2024-02-07 11:53:06 [1054] DEBUG /home/runner/work/nanomq/nanomq/nanomq/pub_handler.c:88: pub_handler: init pipe_info
2024-02-07 11:53:06 [1054] TRACE /home/runner/work/nanomq/nanomq/nng/src/nng.c:390:  ######## nng_ctx_recv context id 4 ######## 
2024-02-07 11:53:06 [1054] TRACE /home/runner/work/nanomq/nanomq/nng/src/sp/protocol/mqtt/nmq_mqtt.c:1003: nano_ctx_recv start 0x636690
2024-02-07 11:53:07 [1089] ERROR /home/runner/work/nanomq/nanomq/nanomq/webhook_inproc.c:201: Connect failed: Timed out
2024-02-07 11:53:11 [1055] ERROR /home/runner/work/nanomq/nanomq/nng/src/sp/protocol/mqtt/auth_http.c:281: Connect failed: Timed out

root@sgw7000:~# gdb nanomq 1045
GNU gdb (GDB) 9.1
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "arm-ostl-linux-gnueabi".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from nanomq...
(No debugging symbols found in nanomq)
Attaching to program: /usr/local/bin/nanomq, process 1045
[New LWP 1050]
[New LWP 1051]
[New LWP 1052]
[New LWP 1053]
[New LWP 1054]
[New LWP 1055]
[New LWP 1056]
[New LWP 1057]
[New LWP 1058]
[New LWP 1059]
[New LWP 1060]
[New LWP 1064]
[New LWP 1065]
[New LWP 1066]
[New LWP 1067]
[New LWP 1068]
[New LWP 1069]
[New LWP 1070]
[New LWP 1071]
[New LWP 1072]
[New LWP 1073]
[New LWP 1074]
[New LWP 1075]
[New LWP 1076]
[New LWP 1077]
[New LWP 1078]
[New LWP 1079]
[New LWP 1080]
[New LWP 1081]
[New LWP 1082]
[New LWP 1083]
[New LWP 1084]
[New LWP 1085]
[New LWP 1086]
[New LWP 1087]
[New LWP 1088]
[New LWP 1089]
[New LWP 1090]
[New LWP 1091]
[New LWP 1092]
[New LWP 1093]
[New LWP 1094]
[New LWP 1095]
[New LWP 1096]
[New LWP 1103]

warning: Unable to find libthread_db matching inferior's thread library, thread debugging will not be available.

warning: Unable to find libthread_db matching inferior's thread library, thread debugging will not be available.
--Type <RET> for more, q to quit, c to continue without paging--
0xb6e85024 in ?? () from /lib/libc.so.6
(gdb) bt
#0  0xb6e85024 in ?? () from /lib/libc.so.6
#1  0xb6edcad8 in ?? () from /lib/libc.so.6
#2  0xb6edcb48 in clock_nanosleep () from /lib/libc.so.6
#3  0xb6ee08be in nanosleep () from /lib/libc.so.6
#4  0x00508f00 in nni_msleep ()
#5  0x0051cec6 in nng_msleep ()
#6  0x004f6bca in broker ()
#7  0x004f79c8 in broker_start ()
#8  0x004f7eee in broker_restart ()
#9  0x004f43e4 in main ()
(gdb) info threads
  Id   Target Id                 Frame 
* 1    LWP 1045 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  2    LWP 1050 "nng:poll:epoll" 0xb6e85026 in ?? () from /lib/libc.so.6
  3    LWP 1051 "nng:resolver"   0xb6f6dc14 in ?? () from /lib/libpthread.so.0
  4    LWP 1052 "nng:task"       0xb6f6dc14 in ?? () from /lib/libpthread.so.0
  5    LWP 1053 "nng:task"       0xb6f6dc14 in ?? () from /lib/libpthread.so.0
  6    LWP 1054 "nng:task"       0xb6f6dc14 in ?? () from /lib/libpthread.so.0
  7    LWP 1055 "nng:task"       0xb6f6dc14 in ?? () from /lib/libpthread.so.0
  8    LWP 1056 "nng:reap2"      0xb6f6dc14 in ?? () from /lib/libpthread.so.0
  9    LWP 1057 "nng:aio:expire" 0xb6f6dc14 in ?? () from /lib/libpthread.so.0
  10   LWP 1058 "nng:aio:expire" 0xb6f6dc14 in ?? () from /lib/libpthread.so.0
  11   LWP 1059 "nng:aio:expire" 0xb6f6dc14 in ?? () from /lib/libpthread.so.0
  12   LWP 1060 "nng:aio:expire" 0xb6f6dc14 in ?? () from /lib/libpthread.so.0
  13   LWP 1064 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  14   LWP 1065 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  15   LWP 1066 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  16   LWP 1067 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  17   LWP 1068 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  18   LWP 1069 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  19   LWP 1070 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  20   LWP 1071 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  21   LWP 1072 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  22   LWP 1073 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
--Type <RET> for more, q to quit, c to continue without paging--
  23   LWP 1074 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  24   LWP 1075 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  25   LWP 1076 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  26   LWP 1077 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  27   LWP 1078 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  28   LWP 1079 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  29   LWP 1080 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  30   LWP 1081 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  31   LWP 1082 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  32   LWP 1083 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  33   LWP 1084 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  34   LWP 1085 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  35   LWP 1086 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  36   LWP 1087 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  37   LWP 1088 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  38   LWP 1089 "nanomq"         0xb6f6dc14 in ?? () from /lib/libpthread.so.0
  39   LWP 1090 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  40   LWP 1091 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  41   LWP 1092 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  42   LWP 1093 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  43   LWP 1094 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  44   LWP 1095 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
  45   LWP 1096 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
--Type <RET> for more, q to quit, c to continue without paging--
  46   LWP 1103 "nanomq"         0xb6e85024 in ?? () from /lib/libc.so.6
(gdb)

**Comment by JaylinYu** [comment]
not seeing anything useful from the log you posted. seems like you pause the program in a breakpoint, that sure gonna stuck.
I need a reproducible client program and the config file you are using. Also would like to verify 2 things:
1. do you enable the Webhook and authentication? could u verify if it is related to your issue by disabling them?
2. Could you test against the 0.20.8 version? I suspect the kicking duplicated clientid breaks sth. But not able to reproduce it on my place.

Basically, if a client sends 2 connect packets within milliseconds in the same TCP connection, the second one gonna be drooped. 
If it sends in 2 separate TCP, then it will be treated as 2 independent session, gonna trigger the session-close due to same client-id.

**Comment by JaylinYu** [comment]
feel free to submit chinese issue if you have troubles.

**Comment by mxchipwanges** [comment]
感谢回复！
你提到的两种情况我们验证后反馈。

另外补充：
1）我们的测试场景是：
一个安卓版的paho.mqtt clien连接nanomq服务器（开启了webhook和http auth）；在大多数单次连接时mqtt, http auth都工作正常，极少数出现上述日志中的connect timeout）。

> 2024-02-07 11:53:07 [1089] ERROR /home/runner/work/nanomq/nanomq/nanomq/webhook_inproc.c:201: Connect failed: Timed out
> 2024-02-07 11:53:11 [1055] ERROR /home/runner/work/nanomq/nanomq/nng/src/sp/protocol/mqtt/auth_http.c:281: Connect failed: Timed out

2）我们反复测试复现，基本确认触发这个问题的原因是安卓系统上没有完全控制好断线重连，经常出现后台反复重连。我们也在敦促安卓客户端改善，但是不同的安卓系统兼容性问题比较多，并且我们实际场景中也确实会存在多设备同时开机发起连接的情况，所以需要保证nanomq server都可以正常被连接。

从日志看每次出现这种情况的时候都是间隔几百毫秒就有两条ssl连接进来，然后出现重复clientid被close，或者其中一条ssl握手失败（mbedtls的问题待查，此时后续重连还能连上），或者这种http auth timeout的异常（后续client都无法连接了）。

**Comment by mxchipwanges** [comment]
配置文件：[nanomq-mx.zip](https://github.com/nanomq/nanomq/files/14203364/nanomq-mx.zip)

**Comment by JaylinYu** [comment]
无论多么频繁的重连都不应该导致broker失去响应。我们自己也测试过很大的重连压力。所以你所说的这种情况不应该出现，若有则是需要优化的地方。
现在需要的是排除其他因素的影响，关闭webhook和auth，并使用旧版本先试试是否有出现吧？

**Comment by mxchipwanges** [comment]
> 无论多么频繁的重连都不应该导致broker失去响应。我们自己也测试过很大的重连压力。所以你所说的这种情况不应该出现，若有则是需要优化的地方。 现在需要的是排除其他因素的影响，关闭webhook和auth，connect timeout是说http请求失败，应不会影响mqtt连接。
> 
> 所以，关闭周边功能 使用旧版本，关闭tls 只使用tcp先试试是否有出现吧？

好的，我们测试后再反馈。

**Comment by mxchipwanges** [comment]
更新：

我们这里反复测试，发现频繁kick重复(mqtt/mqtts) clientid都没有问题；
但是如果开启http auth，(8个client, 间隔500ms)同时重连后，这样很快就会导致nanomq卡住，不响应任何client连接了，也无日志输出了（日志level = trace），只能重启nanomq才行。

测试系统: 
uname -a
Linux mxchip 5.4.31 #1 SMP PREEMPT Fri May 12 07:18:41 UTC 2023 armv7l armv7l armv7l GNU/Linux

cat /proc/cpuinfo 
processor       : 0
model name      : ARMv7 Processor rev 5 (v7l)
BogoMIPS        : 48.00
Features        : half thumb fastmult vfp edsp thumbee neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm 
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x0
CPU part        : 0xc07
CPU revision    : 5

nanomq版本：nanomq-0.21.5-linux-armhf-full.deb
测试client:   mqttx cli
测试命令： mqttx conn -h '192.168.2.3' -p '8883' -u 'mxos' -P '1234' -i 'wes' -k 20 -l mqtts -V 3.1.1 --insecure -rp 500 --maximum-reconnect-times 1000


nanomq错误日志:

2024-03-22 14:27:22 [19986] WARN  /home/runner/work/nanomq/nanomq/nng/src/core/pipe.c:521: *** kick old client pipe(0xa17798) client_id hash(1610094444)  <====这个日志是我测试kick重复clientid时加的。
2024-03-22 14:27:23 [19986] WARN  /home/runner/work/nanomq/nanomq/nng/src/core/pipe.c:521: *** kick old client pipe(0x9fd2ddc0) client_id hash(1610094444)
2024-03-22 14:27:23 [19989] WARN  /home/runner/work/nanomq/nanomq/nng/src/core/pipe.c:521: *** kick old client pipe(0x9fd31208) client_id hash(1610094444)

[同上，省略]

2024-03-22 14:27:44 [19988] WARN  /home/runner/work/nanomq/nanomq/nng/src/core/pipe.c:521: *** kick old client pipe(0x9fe849e8) client_id hash(1610094444)
2024-03-22 14:27:44 [20013] ERROR /home/runner/work/nanomq/nanomq/nanomq/webhook_inproc.c:200: Connect failed: Timed out
2024-03-22 14:27:54 [19987] ERROR /home/runner/work/nanomq/nanomq/nng/src/sp/protocol/mqtt/auth_http.c:312: Read response: Timed out  <===== 有时候错误是  Connect failed: Timed out


^C
root@mxchip :~# tail -f /usr/local/nanomq.log |grep -E "ERROR|kick"
2024-03-22 14:27:44 [19988] WARN  /home/runner/work/nanomq/nanomq/nng/src/core/pipe.c:521: *** kick old client pipe(0x9fe849e8) client_id hash(1610094444)
2024-03-22 14:27:44 [20013] ERROR /home/runner/work/nanomq/nanomq/nanomq/webhook_inproc.c:200: Connect failed: Timed out
2024-03-22 14:27:54 [19987] ERROR /home/runner/work/nanomq/nanomq/nng/src/sp/protocol/mqtt/auth_http.c:312: Read response: Timed out
<====此后client无法再连接，也无日志输出。

**Comment by mxchipwanges** [comment]
> 更新：
> 
> 我们这里反复测试，发现频繁kick重复(mqtt/mqtts) clientid都没有问题； 但是如果开启http auth，(8个client, 间隔500ms)同时重连后，这样很快就会导致nanomq卡住，不响应任何client连接了，也无日志输出了（日志level = trace），只能重启nanomq才行。
> 
> 测试系统: uname -a Linux mxchip 5.4.31 #1 SMP PREEMPT Fri May 12 07:18:41 UTC 2023 armv7l armv7l armv7l GNU/Linux
> 
> cat /proc/cpuinfo processor : 0 model name : ARMv7 Processor rev 5 (v7l) BogoMIPS : 48.00 Features : half thumb fastmult vfp edsp thumbee neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm CPU implementer : 0x41 CPU architecture: 7 CPU variant : 0x0 CPU part : 0xc07 CPU revision : 5
> 
> nanomq版本：nanomq-0.21.5-linux-armhf-full.deb 测试client: mqttx cli 测试命令： mqttx conn -h '192.168.2.3' -p '8883' -u 'mxos' -P '1234' -i 'wes' -k 20 -l mqtts -V 3.1.1 --insecure -rp 500 --maximum-reconnect-times 1000
> 
> nanomq错误日志:
> 
> 2024-03-22 14:27:22 [19986] WARN /home/runner/work/nanomq/nanomq/nng/src/core/pipe.c:521: *** kick old client pipe(0xa17798) client_id hash(1610094444) <====这个日志是我测试kick重复clientid时加的。 2024-03-22 14:27:23 [19986] WARN /home/runner/work/nanomq/nanomq/nng/src/core/pipe.c:521: *** kick old client pipe(0x9fd2ddc0) client_id hash(1610094444) 2024-03-22 14:27:23 [19989] WARN /home/runner/work/nanomq/nanomq/nng/src/core/pipe.c:521: *** kick old client pipe(0x9fd31208) client_id hash(1610094444)
> 
> [同上，省略]
> 
> 2024-03-22 14:27:44 [19988] WARN /home/runner/work/nanomq/nanomq/nng/src/core/pipe.c:521: *** kick old client pipe(0x9fe849e8) client_id hash(1610094444) 2024-03-22 14:27:44 [20013] ERROR /home/runner/work/nanomq/nanomq/nanomq/webhook_inproc.c:200: Connect failed: Timed out 2024-03-22 14:27:54 [19987] ERROR /home/runner/work/nanomq/nanomq/nng/src/sp/protocol/mqtt/auth_http.c:312: Read response: Timed out <===== 有时候错误是 Connect failed: Timed out
> 
> ^C root@mxchip :~# tail -f /usr/local/nanomq.log |grep -E "ERROR|kick" 2024-03-22 14:27:44 [19988] WARN /home/runner/work/nanomq/nanomq/nng/src/core/pipe.c:521: *** kick old client pipe(0x9fe849e8) client_id hash(1610094444) 2024-03-22 14:27:44 [20013] ERROR /home/runner/work/nanomq/nanomq/nanomq/webhook_inproc.c:200: Connect failed: Timed out 2024-03-22 14:27:54 [19987] ERROR /home/runner/work/nanomq/nanomq/nng/src/sp/protocol/mqtt/auth_http.c:312: Read response: Timed out <====此后client无法再连接，也无日志输出。

<=== 补充：client使用非tls也是一样会触发问题。

**Comment by mxchipwanges** [comment]
补充：安装原始版本 nanomq-0.21.5-linux-armhf-full.deb 的测试：

配置文件
[nanomq.conf.txt](https://github.com/nanomq/nanomq/files/14718762/nanomq.conf.txt)

client命令: 
mqttx conn -h '192.168.2.3' -p 1883 -u 'aos' -P '1874' -i 'macos-tcp' -k 20  -l  mqtt -V 3.1.1 --insecure -rp 500 --maximum-reconnect-times 1000

失败日志：
[nanomq_0322_http_timeout.log](https://github.com/nanomq/nanomq/files/14718756/nanomq_0322_http_timeout.log)

**Comment by mxchipwanges** [comment]
> HTTP Auth 是个同步阻塞操作（否则无法保证auth安全）如果都是Read response: Timed out， 说明都是在等待对端的http结果返回，大量的客户端连接耗尽所有taskq线程自然无法继续工作（我猜你的taskq就是8 ），需要等待原来的http请求返回才有余力。这部分目前的设计如此。

目前我们测试异常后，停止所有client请求，nanomq一直无法恢复，只能重启nanomq。
我们希望是停止client请求一段时间后nanomq能够恢复响应，能否优化配置？ 还是说只能外部重启nanomq ？

目前配置文件：taskq 都是配的0

> system {
> 	# # num_taskq_thread
> 	# # Use a specified number of taskq threads 
> 	# #
> 	# # Value: 1-255, Obtain automatically if 0
> 	num_taskq_thread = 0
> 	
> 	# # max_taskq_thread
> 	# # Use a specified maximunm number of taskq threads
> 	# #
> 	# # Value: 1-255, Obtain automatically if 0
> 	max_taskq_thread = 0
> 	
> 	# # parallel
> 	# # Handle a specified maximum number of outstanding requests
> 	# #
> 	# # Value: 1-255, Obtain automatically if 0
> 	parallel = 0
> }

HTTP:

> auth {
> 	# # anonymous
> 	# # allow anonymous login
> 	# #
> 	# # Hot updatable
> 	# # Value: true | false
> 	allow_anonymous = true
> 	
> 	# # Allow or deny if no ACL rules matched.
> 	# #
> 	# # Value: allow | deny
> 	no_match = allow
> 	
> 	# # The action when acl check reject current operation
> 	# #
> 	# # Value: ignore | disconnect
> 	# # Default: ignore
> 	deny_action = ignore
> 	
> 	cache = {
> 		# # The maximum count of ACL entries can be cached for a client.
> 		# #
> 		# # Value: Integer greater than 0
> 		# # Default: 32
> 		max_size = 32
> 		
> 		# # The time after which an ACL cache entry will be deleted
> 		# #
> 		# # Value: Duration
> 		# # Default: 1 minute
> 		ttl = 1m
> 	}
> 	
> 	# # This is password conf file.
> 	# #
> 	# # Value: path string
> 	# # Default: "/etc/pwd.conf"
> 	# password = {include "/etc/nanomq_pwd.conf"}
> 	
> 	# # This is acl conf file.
> 	# #
> 	# # Value: path string
> 	# # Default: "/etc/acl.conf"
> 	# acl = {include "/etc/nanomq_acl.conf"}
> 	
> 	http_auth = {
> 		auth_req {
> 			url = "http://127.0.0.1:8888/mqtt/auth"
> 			method = POST
> 			headers.content-type = "application/json"
> 			params = {clientid = "%c", username = "%u", password = "%P"}
> 		}
> 
> 		super_req {
> 			url = "http://127.0.0.1:8888/mqtt/auth"
> 			method = "POST"
> 			headers.content-type = "application/json"
> 			params = {clientid = "%c", username = "%u", password = "%P"}
> 		}
> 		# # HTTP ACL request is unsupported.
> 		acl_req {
> 			url = "http://127.0.0.1:8888/mqtt/acl"
> 			method = "post"
> 			headers.content-type = "application/x-www-form-urlencoded"
> 			params = {clientid = "%c", username = "%u", access = "%A", ipaddr = "%a", topic = "%t", mountpoint = "%m"}	
> 		}
> 		
> 		# # Time-out time for the request.
> 		timeout = 10s
> 		# # Connection time-out time, used during the initial request,
> 		connect_timeout = 10s
> 		# # Connection process pool size
> 		pool_size = 32
> 	}
> }

**Comment by mxchipwanges** [comment]
更新：
修改了配置如下，暂时16 client重连无异常，只是cpu占用高（单核超100%，双核60%，正常单核8%左右）。
system {
  num_taskq_thread = 32
  max_taskq_thread = 32
  parallel = 32
}

**Comment by JaylinYu** [comment]
那按照设计，aio会有timeout时限，最终应会释放taskq。若一直无法恢复则不应该，会排期优化。
你可以通过增加taskq线程数量来暂时避免工作线程全部耗尽。

**Comment by JaylinYu** [comment]
测试并没有http 阻塞的情况。除非对端服务器一直阻塞不返回。

**Comment by JaylinYu** [comment]
Already refactor http for this
